mayastor-test/test_rebuild.js

```diff
@@ -0,0 +1,379 @@
+        if (err) return reject(err);
+        resolve(data);
+      });
+    });
+  };
+
+  const rebuildState = args => {
+    return new Promise((resolve, reject) => {
+      client.getRebuildState(args, (err, data) => {
+        if (err) return reject(err);
+        resolve(data);
+      });
+    });
+  };
+
+  before(done => {
+    client = createGrpcClient('MayaStor');
+    if (!client) {
+      return done(new Error('Failed to initialize grpc client'));
+    }
+
+    async.series(
+      [
+        common.ensureNbdWritable,
+        next => {
+          fs.writeFile(child1, '', next);
+        },
+        next => {
+          fs.truncate(child1, diskSize, next);
+        },
+        next => {
+          fs.writeFile(child2, '', next);
+        },
+        next => {
+          fs.truncate(child2, diskSize, next);
+        },
+        next => {
+          common.startMayastor(configNexus, ['-r', common.SOCK, '-s', 386]);
+          common.startMayastorGrpc();
+          common.waitFor(pingDone => {
+            // use harmless method to test if the mayastor is up and running
+            client.listPools({}, pingDone);
+          }, next);
+        },
+        next => {
+          createNexus(nexusArgs)
+            .then(() => {
+              next();
+            })
+            .catch(err => {
+              assert(err);
+            })
+            .catch(done);
+        },
+      ],
+      done
+    );
+  });
+
+  after(done => {
+    async.series(
+      [
+        common.stopAll,
+        common.restoreNbdPerms,
+        next => {
+          fs.unlink(child1, err => next());
+        },
+        next => {
+          fs.unlink(child2, err => next());
+        },
+        next => {
+          destroyNexus({ uuid: UUID })
+            .then(() => {
+              next();
+            })
+            .catch(err => {
+              done();
+            })
+            .catch(done);
+        },
+      ],
+      err => {
+        if (client != null) {
+          client.close();
+        }
+        done(err);
+      }
+    );
+  });
+
+  describe('running rebuild', function() {
+    beforeEach(async () => {
+      await addChild(rebuildArgs);
+      await startRebuild(rebuildArgs);
+    });
+
+    afterEach(async () => {
+      await stopRebuild(rebuildArgs);
+      await removeChild(rebuildArgs);
+    });
+
+    it('check nexus state', () => {
+      checkState(ObjectType.NEXUS, 'degraded');
+    });
+
+    it('check source state', () => {
+      checkState(ObjectType.SOURCE_CHILD, 'open');
+    });
+
+    it('check destination state', () => {
+      checkState(ObjectType.DESTINATION_CHILD, 'faulted');
+    });
+  });
+
+  describe('stopping rebuild', function() {
+    beforeEach(async () => {
+      await addChild(rebuildArgs);
+      await startRebuild(rebuildArgs);
+      await stopRebuild(rebuildArgs);
+      // TODO: Check for rebuild stop rather than sleeping
+      await sleep(1000); // Give time for the rebuild to stop
```

Comment:
Should not the stopRebuild return only after the rebuild was stopped? Just like startRebuild returns once rebuild has started. That would eliminate the need for the sleep at this line. Though I'm not sure if it isn't a bad idea from the implementation perspective.

Reply:
At the moment the stopRebuild is an asynchronous call which sets a "stop" flag. It doesn't actually wait for the stop to complete. This is because, in the future, there may be a number of segments rebuilding and we cannot stop the process until these complete. Therefore we didn't want to block the rpc call during this time.